core/src/main/scala/org/apache/spark/Accumulators.scala

```diff
@@ -314,6 +409,59 @@ object AccumulatorParam {
   // TODO: Add AccumulatorParams for other types, e.g. lists and strings
 }
 
+/**
+ * A consistent wrapper of [[org.apache.spark.AccumulatorParam]] where we keep track of
+ * RDD/partitions which have already been processed
+ *
+ * @tparam T type of value to accumulate
+ */
+class ConsistentAccumulatorParam[T](accumulatorParam: AccumulatorParam[T])
+    extends AccumulableParam[UpdateTracking[T], (UpdateInfo, T)] {
+
+  /**
+   * Add additional value to the current accumulator. No consistency checking is perford at this
+   * stage as addAccumulator may be called multiple times inside
+   * the partition.
+   *
+   * @param r the current value of the accumulator
+   * @param t the data to be added to the accumulator
+   * @return the new value of the accumulator
+   */
+  def addAccumulator(r: UpdateTracking[T], t: (UpdateInfo, T)): UpdateTracking[T] = {
+    val v = r.pending.get(t._1).map(
+      accumulatorParam.addAccumulator(_, t._2)
+    ).getOrElse(t._2)
+    r.pending(t._1) = v
+    // We don't need to set the value here since we always have a merge with the zero UpdateTracking
+    r
+  }
+
+  /**
+   * Merge pending updates into the current accumulator.
+   * Checks to make sure that each pending update has not
+   * already been processed before updating.
+   *
+   * @param r1 local source of accumulated
+   * @param r2 set pending updates
+   * @return both data sets merged together
+   */
+  def addInPlace(r1: UpdateTracking[T], r2: UpdateTracking[T]): UpdateTracking[T] = {
```

Comment:
Here you are accumulating r2.pending values to r1. Does r1 also contains pending values or not ? 
I think if it also contains pending values, then it is going to be accumulated if r1 accumulated to other UpdateTracking object, but the last r1 pending values will not be accumulated ever !!